### DISCUSSION TITLE
What is the best way to run data api builder as a windows services ?

### DISCUSSION THREAD

**Discussion by diallonortv** [Azure/data-api-builder#1625] [General]
Hello,

Which direction should I go to run the data-api-builder as a Windows service ( windows server 2022 ) ? I only have access to VMWARE virtual machine. Therefore, a container is not really a good option for me.

Any pieces of information is greatly appreciated.

Regards

Vincent

**Comment by seantleonard** [comment]
Depending on what you're allowed to configure on your WS2022/VMWare instance, you can reference this collection of docs describing how to host ASP.NET apps: https://learn.microsoft.com/aspnet/core/host-and-deploy/?view=aspnetcore-6.0

More details on possible options in the section [Hosting Models](https://learn.microsoft.com/aspnet/core/fundamentals/servers/?view=aspnetcore-6.0&tabs=windows#hosting-models)

The easiest route would be Kestrel self-hosting, and you'd need to set up HTTPS accordingly.

> Several hosting models are available:
> 
> Kestrel self-hosting: The Kestrel web server runs without requiring any other external web server such as IIS or HTTP.sys.
> 
> HTTP.sys self-hosting is an alternative to Kestrel. Kestrel is recommended over HTTP.sys unless the app requires features not available in Kestrel.
> 
> IIS in-process hosting: An ASP.NET Core app runs in the same process as its IIS worker process. IIS in-process hosting provides improved performance over IIS out-of-process hosting because requests aren't proxied over the loopback adapter, a network interface that returns outgoing network traffic back to the same machine. IIS handles process management with the [Windows Process Activation Service (WAS)](https://learn.microsoft.com/en-us/iis/manage/provisioning-and-managing-iis/features-of-the-windows-process-activation-service-was).
> 
> IIS out-of-process hosting: ASP.NET Core apps run in a process separate from the IIS worker process, and the module handles process management. The module starts the process for the ASP.NET Core app when the first request arrives and restarts the app if it shuts down or crashes. This is essentially the same behavior as seen with apps that run in-process that are managed by the [Windows Process Activation Service (WAS)](https://learn.microsoft.com/en-us/iis/manage/provisioning-and-managing-iis/features-of-the-windows-process-activation-service-was). Using a separate process also enables hosting more than one app from the same app pool.

**Reply by diallonortv** [reply]
Thanks a lot
